src/consensus.rs

```diff
@@ -1,107 +1,139 @@
-            .unbounded_send(u.into())
-            .expect("Channel to extender should be open.")
-    }));
-
-    let (terminal_exit, exit_rx) = oneshot::channel();
-    let mut terminal_handle = spawn_handle
-        .spawn_essential(
-            "consensus/terminal",
-            async move { terminal.run(exit_rx).await },
-        )
-        .fuse();
-    info!(target: "AlephBFT", "{:?} All services started.", conf.node_ix);
-
-    let mut terminal_exited = false;
-    let mut creator_exited = false;
-    let mut extender_exited = false;
-    futures::select! {
-        _ = exit => {},
-        _ = terminal_handle => {
-            terminal_exited = true;
-            debug!(target: "AlephBFT-consensus", "{:?} terminal task terminated early.", conf.node_ix);
-        },
-        _ = creator_handle => {
-            creator_exited = true;
-            debug!(target: "AlephBFT-consensus", "{:?} creator task terminated early.", conf.node_ix);
-        },
-        _ = extender_handle => {
-            extender_exited = true;
-            debug!(target: "AlephBFT-consensus", "{:?} extender task terminated early.", conf.node_ix);
+}
+
+impl<H: Hasher, SH: SpawnHandle> Consensus<H, SH> {
+    pub(crate) fn new(
+        conf: Config,
+        spawn_handle: SH,
+        incoming_notifications: Receiver<NotificationIn<H>>,
+        outgoing_notifications: Sender<NotificationOut<H>>,
+        ordered_batch_tx: Sender<OrderedBatch<H::Hash>>,
+    ) -> Self {
+        Consensus {
+            conf,
+            spawn_handle,
+            incoming_notifications,
+            outgoing_notifications,
+            ordered_batch_tx,
         }
     }
 
-    // we stop no matter if received Ok or Err
-    let _ = terminal_exit.send(());
-    if !terminal_exited {
-        terminal_handle.await.unwrap();
-    }
-    let _ = creator_exit.send(());
-    if !creator_exited {
-        creator_handle.await.unwrap();
-    }
-    let _ = extender_exit.send(());
-    if !extender_exited {
-        extender_handle.await.unwrap();
-    }
+    pub(crate) async fn run(self, mut exit: oneshot::Receiver<()>) {
+        info!(target: "AlephBFT", "{:?} Starting all services...", self.conf.node_ix);
+
+        let n_members = self.conf.n_members;
+
+        let (electors_tx, electors_rx) = mpsc::unbounded();
+        let mut extender = Extender::<H>::new(
+            self.conf.node_ix,
+            n_members,
+            electors_rx,
+            self.ordered_batch_tx,
+        );
+        let mut barrier = Barrier::new();
+
+        let extender_barrier = barrier.clone("consensus/extender");
+        let (extender_exit, exit_rx) = oneshot::channel();
+        let extender_handle = self
+            .spawn_handle
+            .spawn_essential("consensus/extender", async move {
+                extender.extend(exit_rx).await;
+
+                extender_barrier.wait().await;
+            });
+        let mut extender_handle = into_infinite_stream(extender_handle).fuse();
```

Comment:
Yuck! Converting it to a stream is so artificial, given that we will poll at most two elements... How about making it a fused future and replacing  `alerter_handle.next().await.unwrap()` at the end of run with something like 
`alerter_handle.now_or_never().unwrap_or(())`?

Reply:
I agree it is `yuck`, but just pretty convenient. I will try to think of some cleaner solution.

Reply:
Is thar aything wrong with the solution I proposed?
```
let extender_handle = extender_handle.fuse();
...
select! {
    ...
}
...
extender_handle.now_or_never().unwrap_or(())
```

Reply:
> Evaluates and consumes the future, returning the resulting output if the future is ready after the first call to Future::poll. If poll instead returns Poll::Pending, None is returned. This method is useful in cases where immediacy is more important than waiting for a result. It is also convenient for quickly obtaining the value of a future that is known to always resolve immediately.

My understanding is: if that `extender_handle` didn't return `Poll::Ready()` inside of that `select` statement, then that `extender_handle.now_or_never().unwrap_or(())` will not wait till it returns `Poll::Ready` and we want to exit from this method after we know that extender already finished.